gremlin-go/driver/graph.go

```diff
@@ -120,3 +121,63 @@ func (p *Path) GetPathObject(key string) (interface{}, error) {
 		return nil, fmt.Errorf("path does not contain a label of '%s'", key)
 	}
 }
+
+// SetTemplate describes the necessary methods that need to be implemented to make a user-defined Set implementation.
+// Set is the default implementation that should be used but SetTemplate can be used for a custom implementation if needed.
+type SetTemplate interface {
+	convertSetToSlice() ([]interface{}, error)
+	convertSliceToSet([]interface{}) (SetTemplate, error)
+}
+
+// Set is a default custom declaration since Go does not natively provide this feature.
+type Set struct {
+	Objects []interface{}
+}
+
+func (s *Set) convertSetToSlice() ([]interface{}, error) {
+	return s.Objects, nil
+}
+
+func (s *Set) convertSliceToSet(in []interface{}) (SetTemplate, error) {
+	slice, err := interfaceToSlice(in)
+	if err != nil {
+		return nil, err
+	}
+	s.Objects = slice
+	return s, nil
+}
+
+// Construct a new Set with filtered values
+func newSet(in interface{}) *Set {
+	slice, _ := interfaceToSlice(in)
+	filtered := removeDuplicateValues(slice)
+	s := new(Set)
+	s.Objects = filtered
+	return s
+}
+
+func interfaceToSlice(in interface{}) ([]interface{}, error) {
+	var interfaceSlice []interface{}
+	switch reflect.TypeOf(in).Kind() {
+	case reflect.Slice:
+		slice := reflect.ValueOf(in)
+		for i := 0; i < slice.Len(); i++ {
+			interfaceSlice = append(interfaceSlice, slice.Index(i).Interface())
+		}
+	default:
+		return nil, errors.New("input is not a slice")
+	}
+	return interfaceSlice, nil
+}
+
+func removeDuplicateValues(slice []interface{}) []interface{} {
```

Comment:
Instead of removing duplicates after they are added, just check for them when we're adding.

Reply:
agree